src/gc-debug.c

```diff
@@ -250,6 +253,120 @@ void gc_verify(jl_ptls_t ptls)
 }
 #endif
 
+#ifdef MEMFENCE
+static uint8_t freelist_map[GC_PAGE_SZ / sizeof(void*) / 8];
+static int freelist_zerod;
+void gc_verify_tags(void)
+{
+    // verify the freelist chains look valid
+    for (int t_i = 0; t_i < jl_n_threads; t_i++) {
+        jl_ptls_t ptls2 = jl_all_tls_states[t_i];
+        for (int i = 0; i < JL_GC_N_POOLS; i++) {
+            // for all pools, iterate its freelist
+            jl_gc_pool_t *p = &ptls2->heap.norm_pools[i];
+            jl_taggedvalue_t *next = p->freelist;
+            jl_taggedvalue_t *last = NULL;
+            char *allocating = gc_page_data(next);
+            while (next) {
+                // and assert that the freelist values aren't gc-marked
+                assert(next->bits.gc == 0);
+                // TODO: verify they are ordered and on the right byte boundaries
+                if (gc_page_data(next) != gc_page_data(last)) {
+                    // and verify that the chain looks valid
+                    jl_gc_pagemeta_t *pg = page_metadata(next);
+                    assert(last == NULL || pg->osize == page_metadata(last)->osize);
+                    if (gc_page_data(next) != allocating) {
+                        // when not currently allocating on this page, fl_begin_offset should be correct
+                        assert(next == page_pfl_beg(pg));
+                    }
+                }
+                last = next;
+                next = next->next;
+            }
+        }
+    }
+
+
+    // verify that all the objects on every page are either valid julia objects
+    // or are part of the freelist or are on the allocated half of a page
+    for (int region_i = 0; region_i < REGION_COUNT; region_i++) {
+        if (!regions[region_i].pages)
+            break;
+        region_t *region = &regions[region_i];
+        for (int pg_i = 0; pg_i <= region->ub; pg_i++) {
+            uint32_t line = region->allocmap[pg_i];
+            if (line) {
+                for (int j = 0; j < 32; j++) {
+                    if ((line >> j) & 1) {
+                        // for all pages in use
+                        jl_gc_pagemeta_t *pg = &region->meta[pg_i*32 + j];
+                        int p_n = pg->pool_n;
+                        int t_n = pg->thread_n;
+                        jl_ptls_t ptls2 = jl_all_tls_states[t_n];
+                        jl_gc_pool_t *p = &ptls2->heap.norm_pools[p_n];
+                        int osize = pg->osize;
+                        char *data = pg->data;
+                        char *page_begin = data + GC_PAGE_OFFSET;
+                        jl_taggedvalue_t *v = (jl_taggedvalue_t*)(data + GC_PAGE_OFFSET);
+                        char *lim = (char*)v + GC_PAGE_SZ - GC_PAGE_OFFSET - osize;
+                        // reset the freelist map to zero
+                        if (!freelist_zerod) {
```

Comment:
Instead of using a global buffer, just do `uint8_t freelist_map[...] = {};`? It's at most 512 bytes and zeroing it is probably cheap enough compare to other operations, especially for a debug option.